README.md
# Tranalator
基于python语言写的文档翻译器，可实现word文档翻译，突破了百度翻译和google翻译只能一次翻译5000个单词的限制。


changelog.md
2018/4/27 创建代码仓库
2018/4/28 更新了只能谷歌翻译的问题


Fanvenapp/Translate/calcTk.py
# /usr/bin/env python
# -*- coding: utf-8 -*-
# @Date    : 2018-04-27 09:49:44
# @Author  : HannochTao (hannochtao@163.com)
# @Link    : http://www.imstudy.online
# @Version : 1.4

import execjs

class Py4Js():
    
    def __init__(self):
        self.ctx = execjs.compile("""
        function TL(a) {
        var k = "";
        var b = 406644;
        var b1 = 3293161072;
        
        var jd = ".";
        var $b = "+-a^+6";
        var Zb = "+-3^+b+-f";
    
        for (var e = [], f = 0, g = 0; g < a.length; g++) {
            var m = a.charCodeAt(g);
            128 > m ? e[f++] = m : (2048 > m ? e[f++] = m >> 6 | 192 : (55296 == (m & 64512) && g + 1 < a.length && 56320 == (a.charCodeAt(g + 1) & 64512) ? (m = 65536 + ((m & 1023) << 10) + (a.charCodeAt(++g) & 1023),
            e[f++] = m >> 18 | 240,
            e[f++] = m >> 12 & 63 | 128) : e[f++] = m >> 12 | 224,
            e[f++] = m >> 6 & 63 | 128),
            e[f++] = m & 63 | 128)
        }
        a = b;
        for (f = 0; f < e.length; f++) a += e[f],
        a = RL(a, $b);
        a = RL(a, Zb);
        a ^= b1 || 0;
        0 > a && (a = (a & 2147483647) + 2147483648);
        a %= 1E6;
        return a.toString() + jd + (a ^ b)
    };
    
    function RL(a, b) {
        var t = "a";
        var Yb = "+";
        for (var c = 0; c < b.length - 2; c += 3) {
            var d = b.charAt(c + 2),
            d = d >= t ? d.charCodeAt(0) - 87 : Number(d),
            d = b.charAt(c + 1) == Yb ? a >>> d: a << d;
            a = b.charAt(c) == Yb ? a + d & 4294967295 : a ^ d
        }
        return a
    }
    """)
        
    def getTk(self,text):
        return self.ctx.call("TL",text)


Fanvenapp/Translate/baidutranslate.py
# /usr/bin/env python
# -*- coding: utf-8 -*-
# @Date    : 2018-04-27 09:49:44
# @Author  : HannochTao (hannochtao@163.com)
# @Link    : http://www.imstudy.online
# @Version : 1.4

from requests import Session
from Translate import baidutransapi
import re

__version__ = "v1.2"
__baidulanguage__ = {'英语': 'en', '中文': 'zh', '日语': 'jp', 'auto': 'en'}

s = Session()


def baidu_translate(text: object = '', to: object = '') -> object:
    """
    :param text: type content
    :param to: auto Chinese
    :return: translate result
    """
    global html

    try:
        baidu_transapi_data = {
            'from': 'auto',
            'to': __baidulanguage__[to],
            'query': text,
            'source': 'txt'
        }
    except KeyError:
        baidu_transapi_data = {
            'from': 'auto',
            'to': __baidulanguage__['auto'],
            'query': text,
            'source': 'txt'
        }

    count = 20
    while count:
        try:
            html = s.post(baidutransapi.baidu_transapi, data=baidu_transapi_data, timeout=2)
        except TimeoutError:
            count -= 1
            continue
        break

    if count == 0:
        return '您当前的环境网络不稳定，建议到网络良好的环境下使用！'

    try:
        html.json()['data']
    except (IndexError, ValueError, KeyError):
        try:
            html.json()['result']
        except (IndexError, ValueError, KeyError):
            return '翻译失败！'
        else:
            result = html.json()['result']

            r = re.compile(r'.*?"voice":\[{.*?"(\[.+?])"}]')
            voice = r.findall(result)
            try:
                voice[0]
            except:
                pass
            finally:
                phonic = ''
                for i in range(2):
                    try:
                        phonic += voice[i]
                        phonic += ' '
                    except:
                        break
                phonic = phonic.replace('"phonic"', '').replace('"en_phonic"', '').replace('"us_phonic"', '') \
                    .replace('"', '').replace(':', '').replace('{', '').replace('}', '') \
                    .replace(',', ' ')

                r = re.compile(r'{"src".*?"mean":.*?{(.+?)}}')
                results = r.findall(result)
                result = ''
                for i in range(31):
                    try:
                        result += results[i]
                    except:
                        break
                result = result.replace('"cont"', '').replace('"pre"', '').replace('"', '').replace(':', '') \
                    .replace('{', '').replace('}', '').replace('1,', '; ').replace('0,', '; ').replace(',', '') \
                    .replace('1', '').replace('0', '')
                return phonic + '\r\n' + result
    else:
        results = []
        for i in range(1000):
            try:
                results.append(html.json()['data'][i]['dst'])
            except:
                break
        return str(results).replace('[', '').replace(']', '').replace('\'', '').replace(',', '\r\n')


Fanvenapp/Translate/googletranlate.py
# /usr/bin/env python
# -*- coding: utf-8 -*-
# @Date    : 2018-04-27 09:49:44
# @Author  : HannochTao (hannochtao@163.com)
# @Link    : http://www.imstudy.online
# @Version : 1.4

import requests
from Translate.calcTk import Py4Js
from bs4 import BeautifulSoup
import re

__goolelanguage__ = {'英语': 'en', '中文': 'zh-CN', '日语': 'ja', 'auto': 'en'}


def google_translate(text: object = '', to_language: object = '') -> object:
    """
    :param text: type content
    :param to: auto Chinese
    :return: translate result
    """
    js = Py4Js()
    tk = js.getTk(text)

    header = {'content-type': 'application/json',
              'User-Agent': 'Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:22.0) Gecko/20100101 Firefox/22.0'}
    # 要访问的url类似如下试例，把下面的url分割
    """
    https://translate.google.cn/translate_a/single?client=t&sl=zh-CN&tl=ja&hl=zh-CN&dt=at&dt=bd&dt=ex&dt=ld
    &dt=md&dt=qca&dt=rw&dt=rm&dt=ss&dt=t&ie=UTF-8&oe=UTF-8&source=bh&ssel=0
    &tsel=0&kc=1&tk=440454.52746&q=%E6%88%91%E6%98%AF%E4%B8%AD%E5%9B%BD%E4%BA%BA
    """
    url_one = "http://translate.google.cn/translate_a/single?client=t&sl="
    sl = 'auto'
    tl = '&tl=' + __goolelanguage__[to_language]
    url_two = "&hl=zh-CN&dt=at&dt=bd&dt=ex&dt=ld&dt=md&dt=qca&dt=rw&dt=rm&dt=ss&dt=t&ie=UTF-8&oe=UTF-8&clearbtn=1&otf=1&pc=1&srcrom=0&ssel=0&tsel=0&kc=2&tk="
    token = tk
    q = '&q=' + text
    # 把参数拼接成url
    url = url_one + sl + tl + url_two + token + q
    # print(url)
    response = requests.get(url, headers=header)

    # print(response.text)
    # 解析网页得到翻译结果
    try:
        soup = BeautifulSoup(response.content, 'lxml')
        # print(str(soup))
        tra_result = re.findall('\[\"(.*?)\"\,\"', str(soup))
        str_tra_result = ''.join(tra_result)
        result = str_tra_result.replace(r'\n', '\n')
        return result
        #print(result)
    except:
        print("Translation Failed!")




Fanvenapp/Fanven.py
# /usr/bin/env python
# -*- coding: utf-8 -*-
# @Date    : 2018-04-27 09:49:44
# @Author  : HannochTao (hannochtao@163.com)
# @Link    : http://www.imstudy.online
# @Version : 1.4

from Translate.baidutranslate import baidu_translate, __version__
from Translate.googletranlate import google_translate
from __init__ import __author__, __copyright_info__, __doc__,__help1__,__help2__,__help3__
from tkinter import *
from tkinter.ttk import Combobox
from tkinter.scrolledtext import ScrolledText
from tkinter import messagebox
import  tkinter.filedialog
import docx
import re

class JiroApplication(Frame):


    def __init__(self, master=None):
        Frame.__init__(self, master=master)
        self.pack()
        self.create_window()

    def create_window(self):

        def translate_link(documnt_conent):
            if translate_link_chosen.current(0):
                translate_result = baidu_translate(text=documnt_conent,
                                                   to=language.get())
                # 在翻译结果区域显示结果
                trans_result_box.insert(END, translate_result)
            else:
                translate_result = google_translate(text=documnt_conent,
                                                    to_language=language.get())
                # 在翻译结果区域显示结果
                trans_result_box.insert(END, translate_result)

        def documnt_translate(event):

            documnt_conent = type_content_box.get('1.0', 'end-1c')
            # 统计有多少字符
            # print(type_content_box.count('1.0', END))
            # 匹配五段，以换行为界限
            pattern = re.compile(r'.*?\n.*?\n.*?\n.*?\n.*?\n')
            find_paragraphs = pattern.findall(documnt_conent)

            for find_paragraph in find_paragraphs:
                if type_content_box.count('1.0', END) > tuple([0]):
                    translate_link(documnt_conent)

            finally_pattern = re.compile(find_paragraphs[-1] + '(.*?)')
            finally_paragraph = finally_pattern.search(documnt_conent)
            #print(find_paragraph)
            translate_link(documnt_conent)

        def boxconnent_translate():
            documnt_conent = type_content_box.get('1.0', 'end-1c')
            if type_content_box.count('1.0', END) > tuple([0]):
                translate_link(documnt_conent)



        # 显示翻译结果
        def show_trans_result():
            # 运行结果按钮状态:正在翻译
            run_after.config(text='正在翻译', bg='#ffffff', fg='#000000', )
            trans_result_box.delete(1.0, END)
            # 翻译文档 事件绑定鼠标左键按下放开时，翻译文档
            openfile_button.bind('ButtonRelease-1',documnt_translate)
            #直接翻译文本框里的内容
            boxconnent_translate()
            # 运行结果按钮状态:运行结果
            run_after.config(text='运行结果')

        def get_documnt(file_path):
            # 获取文档对象

            file_path_ = file_path
            file = docx.Document(file_path_)
            #print("段落数:" + str(len(file.paragraphs)))  # 段落数为13，每个回车隔离一段
            # 输出段落编号及段落内容
            type_content_box.delete(1.0,END)
            for i in range(len(file.paragraphs)):
                type_content_box.insert(END, file.paragraphs[i].text)
                type_content_box.insert(END,'\n')


        # 选择文件路径
        def selectPath():
            path_ = askdirectory()
            path.set(path_)

        def file_path():
            filetypes = [
                ("All Files", '*'),
                ("word 97-2003文档", '*.doc', 'TEXT'),
                ("word文档",'*.docx','TEXT')]

            filename = tkinter.filedialog.askopenfilename(filetypes=filetypes)
            if filename != '':
                openfile_lable.config(text="您选择的文件是：" + filename)
                get_documnt(filename)
            else:
                openfile_lable.config(text="您没有选择任何文件")

        def save_file():
            text_value = trans_result_box.get('1.0', END).strip()
            if text_value:
                filename = tkinter.filedialog.asksaveasfile()
                #print(filename.name)
                if filename:
                    filename.write(text_value + '\n')

                savefile_lable.config(text="您选择的文件是：" + filename.name)

        #打开文件标签
        openfile_lable = Label(self.master, text='请选择word文档', width=47, bg='#6495ED', fg='#ffffff',
                                  relief='flat',font=('微软雅黑', 11))
        openfile_lable.place(x=1, y=10)
        #保存文件标签
        savefile_lable = Label(self.master, text='请选择保存路径', width=40, bg='#6495ED', fg='#ffffff',
                             relief='flat', font=('微软雅黑', 11))
        savefile_lable.place(x=510, y=10)
        # 打开文件按钮
        openfile_button = Button(self.master, text="打开文件", command=lambda: file_path())
        openfile_button.place(x=430, y=10)
        #保存按钮
        save_button = Button(self.master, text="保存文件", command=lambda: save_file())
        save_button.place(x=960-80, y=10)

        ###############################################
        # 自动检测语言
        auto_language = Button(self.master, text='    自动检测语言    ', state='disabled', relief='flat',
                               bg='#ffffff', fg='#000000', font=('微软雅黑', 9), )
        auto_language.place(x=1, y=50)

        # 选择语言
        language = StringVar()
        language_chosen = Combobox(self.master, width=10, textvariable=language, )
        language_chosen['values'] = ['中文', '英语', '日语']  # 设置下拉列表的值
        language_chosen.place(x=25 * 7, y=50)
        language_chosen.current(0)

        # 选择百度或者google翻译，默认为百度翻译
        translate_link_chosen = Combobox(self.master, width=10)
        translate_link_chosen['values'] = ['百度翻译', '谷歌翻译']  # 设置下拉列表的值
        translate_link_chosen.place(x=40 * 7, y=50)
        translate_link_chosen.current(0)
        # 翻译按钮
        translate_button = Button(self.master, text='翻   译', width=10, bg='#6495ED', fg='#ffffff',
                                  relief='flat', command=lambda: show_trans_result(),
                                  font=('微软雅黑', 9))
        translate_button.place(x=60 * 7 + 5, y=50)

        # 运行结果按钮
        run_after = Button(self.master, state=DISABLED, width=10, relief='flat', font=('微软雅黑', 9), )
        run_after.place(x=72 * 7 + 4, y=50)

        #########################################################################
        # 在窗口中创建输入子窗口
        type_content_box = ScrolledText(self.master, width=55, height=15, font=('微软雅黑', 11), )
        type_content_box.insert(END, '请在此处输入文字')
        type_content_box.focus_set()
        type_content_box.pack(side=LEFT, expand=True)

        # 结果窗口
        trans_result_box = ScrolledText(self.master, width=55, height=15, font=('consolas', 11), )
        trans_result_box.pack(side=RIGHT)

        ################################################
        #帮助
        help1 = Label(self.master, text=__help1__, font=('微软雅黑', 11), )
        help1.place(x=1, y = 480 - 90)

        help2 = Label(self.master, text=__help2__, font=('微软雅黑', 11), )
        help2.place(x=1, y=480 - 70)

        help3 = Label(self.master, text=__help3__, font=('微软雅黑', 11), )
        help3.place(x=1, y=480 - 50)

        # 底部版权信息
        copyright = Label(self.master, text=__copyright_info__, font=('consolas', 11), )
        copyright.place(x=(960 - len(__copyright_info__) * 7.5) / 2,
                        y=480 - 22)


WIDTH = 960
HEIGHT = 480

app_name = 'Fanven ' + __version__

root = Tk()

# 隐藏窗口
root.withdraw()

# 窗口标题
root.title(app_name)

# 设置窗口透明度 第二个参数数值越小则越透明
root.attributes('-alpha', 0.90)

# 窗口大小、居中
root.geometry('{}x{}+{}+{}'.format(str(WIDTH), str(HEIGHT),
                                   str(int((root.winfo_screenwidth() - WIDTH) / 2)),
                                   str(int((root.winfo_screenheight() - HEIGHT) / 2))))

# 固定窗口大小
root.resizable(0, 0)

# 弹出对话框
#messagebox.showinfo(title='Fanven Box', message=__doc__ + '\n\t\t\t————' + __author__)
# 显示窗口
root.deiconify()

app = JiroApplication(root)

app.mainloop()
